server/server.go

```diff
@@ -6,53 +6,287 @@ package server
+	defer stop()
 
+	// before reading notification we send the current values
+	// in case the notification would not come in a long time
+	err = producer.Send(&messages.PersistedIndexReply{
+		Uuid:           serv.uuid,
+		PersistedIndex: serv.GetPersistedIndex(),
+	})
 	if err != nil {
-		return fmt.Errorf("error in StreamAcknowledgements: %w", err)
+		return err
 	}
+
+	for {
+		select {
+
+		case change, open := <-ch:
+			if !open || change.persistedIndex == nil {
+				continue
+			}
+
+			err := producer.Send(&messages.PersistedIndexReply{
+				Uuid:           serv.uuid,
+				PersistedIndex: *change.persistedIndex,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to send the update: %w", err)
+			}
+
+		case <-producer.Context().Done():
+			return fmt.Errorf("producer context: %w", producer.Context().Err())
+
+		case <-serv.polling.ctx.Done():
+			return fmt.Errorf("server is stopped: %w", serv.polling.ctx.Err())
+		}
+	}
+}
+
+// Close implements the Closer interface
+func (serv *shipperServer) Close() error {
+	serv.close.Do(func() {
+		// polling must be stopped first, otherwise it would try to write
+		// a notification to a closed channel and this would cause a panic
+		serv.polling.stop()
+		<-serv.polling.stopped
+
+		serv.logger.Debug("shutting down all notifications...")
+		serv.notifications.shutdown()
+		serv.logger.Debug("all notifications have been shut down")
+	})
+
+	return nil
+}
+
+func (serv *shipperServer) startPolling() {
+	go func() {
+		ticker := time.NewTicker(serv.cfg.PollingInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+
+			case <-serv.polling.ctx.Done():
+				err := serv.polling.ctx.Err()
+				if err != nil && errors.Is(err, context.Canceled) {
+					serv.logger.Error(err)
+				}
+				close(serv.polling.stopped) // signaling back to `Close`
+				return
+
+			case <-ticker.C:
+				serv.logger.Debug("updating indices...")
+				err := serv.updateIndices(serv.polling.ctx)
+				if err != nil {
+					serv.logger.Errorf("failed to update indices: %s", err)
+				} else {
+					serv.logger.Debug("successfully updated indices.")
+				}
+			}
+		}
+	}()
+}
+
+// updateIndices updates in-memory indices and notifies subscribers if necessary.
+func (serv *shipperServer) updateIndices(ctx context.Context) error {
+	log := serv.logger
+	c := change{}
+
+	oldPersistedIndex := serv.GetPersistedIndex()
+	persistedIndex := uint64(serv.publisher.PersistedIndex())
+
+	if persistedIndex != oldPersistedIndex {
+		atomic.StoreUint64(&serv.persistedIndex, persistedIndex)
+		// register the change
+		c.persistedIndex = &persistedIndex
+		log = log.With("persisted_index", persistedIndex)
+	}
+
+	if c.Any() {
+		log.Debug("indices have been updated")
+
+		log := log.With(
+			"subscribers_count", len(serv.notifications.subscribers),
+		)
+
+		// this must be async because the loop in `notifyChange` can block on a receiver channel
+		go func() {
+			log.Debug("notifying about change...")
+			serv.notifications.notify(ctx, c)
```

Comment:
If notify can block, what prevents this loop from just indefinitely accumulating goroutines that are blocked in `notify`?

Reply:
Hmm, to me the ideal behaviour would be that the latest index value overwrites any pending notification for an older value. I don't think clients need to see every incremental update, they just need to know the latest state. 

I just don't know how you would achieve this with channels. Maybe you could do this with a [condition variable](https://pkg.go.dev/sync#Cond) on a single element queue, like a channel of size 1 where you can edit the underlying buffer in place.

Reply:
So I was thinking about how to do this with channels (`PersistedIndex` could do some extra work to try and use the most recent value) but instead I'm going to suggest that we can get away without all this synchronization entirely: there's no particular need for all listeners to be updated simultaneously with identical values. All we're doing is sending a periodic watermark. We can remove the listeners and notifications, and just say that every connection via `PersistedIndex` gets its own timer, and when it fires it gets the current value from the queue and sends it if it's changed.

This way there's never concern about a particular listener blocking and missing messages, because it's blocking by itself and when it's done it will fetch the latest value itself, there is nothing to accumulate. And each listener can have its "current" value be a local variable, so no synchronization calls are needed to compare or update it. And calls to fetch the queue state are cheap, so doing it independently for each connection should have pretty comparable performance with a lot less metadata.

Reply:
The problem with using `Cond` is that there is no way to cancel `Wait` when the client decided to canceled its subscription. All the subscription handlers will hang until the next persistent index update.

Technically it's not a big deal if the index is changing quite often. I'll try using `Cond` instead and see if it causes any issues in the tests.

Reply:
@cmacknz @faec please see https://github.com/elastic/elastic-agent-shipper/pull/76/commits/48d073c16d6b41c77cc2dc527b49c39217224769

It's simple, safe and well tested.

Reply:
I'm not sure if that link relates to my response -- my suggestion was to bypass the notifications and condition variables entirely since there's no need for different clients to be notified synchronously about persistent index state. You should be able to handle everything directly from `shipperServer.PersistedIndex` by creating a local timer channel and checking the queue state directly instead of going through `serv.notifications` for synchronous changes.

Reply:
(Basically my suggestion is that we can remove `notifications.go` entirely, as well as ~all shared synchronization variables and cross-routine communication, for the cost of one timer per active connection, which seems like a good trade to me.)

Reply:
@cmacknz 

I thought the whole shipper initiative started because our customers were running thousands of agents and hundreds of inputs each. I remember these numbers told during one of our meetings, can't find a source unfortunately.

What's surprising to me is that we're concerned about performance when there are channel-based subscriptions or use of atomics but we change our estimate when there are timers instead, saying that it does not need to scale much.

Reply:
@cmacknz @faec 

For the record, I'm still against the use of timers and I think it's worse than the current implementation but if the team decides to go forward with this, I'll commit. 

Switching to a potentially less scalable solution just to remove some code is not enough reasoning for me personally. Therefore if we're to address this, I will create benchmarks and compare 2 different solutions.